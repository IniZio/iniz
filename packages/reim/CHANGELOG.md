# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="2.0.2"></a>
## [2.0.2](https://github.com/IniZio/reim/compare/v2.0.1...v2.0.2) (2019-05-13)


### Bug Fixes

* remove lodash from reim ([c3de99b](https://github.com/IniZio/reim/commit/c3de99b))




<a name="2.0.1"></a>
## [2.0.1](https://github.com/IniZio/reim/compare/v2.0.0...v2.0.1) (2019-05-13)


### Bug Fixes

* **react:** unncessary rerender fix ([92e23e7](https://github.com/IniZio/reim/commit/92e23e7))




<a name="2.0.0"></a>
# [2.0.0](https://github.com/IniZio/reim/compare/v1.17.0...v2.0.0) (2019-05-12)


### Bug Fixes

* **plugin:** make plugins work again ([f3512f9](https://github.com/IniZio/reim/commit/f3512f9))
* **State:** Replace getter/setter with filter/actions in State ([0178477](https://github.com/IniZio/reim/commit/0178477))


### Features

* **actions:** add actions and effects methods ([00ebcfb](https://github.com/IniZio/reim/commit/00ebcfb))




<a name="1.17.0"></a>
# [1.17.0](https://github.com/IniZio/reim/compare/v1.16.0...v1.17.0) (2019-01-15)


### Features

* **getter:** allow string as getter ([98c4483](https://github.com/IniZio/reim/commit/98c4483))




<a name="1.16.0"></a>
# [1.16.0](https://github.com/IniZio/reim/compare/v1.15.1...v1.16.0) (2019-01-14)


### Features

* **devtool:** supports redux devtool ([#43](https://github.com/IniZio/reim/issues/43)) ([80a78a2](https://github.com/IniZio/reim/commit/80a78a2)), closes [#40](https://github.com/IniZio/reim/issues/40)




<a name="1.15.1"></a>
## [1.15.1](https://github.com/IniZio/reim/compare/v1.15.0...v1.15.1) (2019-01-14)


### Bug Fixes

* **react-hook:** unnecessary rerender on mount and store update ([#42](https://github.com/IniZio/reim/issues/42)) ([c85e4df](https://github.com/IniZio/reim/commit/c85e4df)), closes [#41](https://github.com/IniZio/reim/issues/41)




<a name="1.13.0"></a>
# [1.13.0](https://github.com/IniZio/reim/compare/v1.12.1...v1.13.0) (2018-12-21)


### Features

* **hook:** adds hook support for react-reim ([#36](https://github.com/IniZio/reim/issues/36)) ([1496771](https://github.com/IniZio/reim/commit/1496771)), closes [#34](https://github.com/IniZio/reim/issues/34)




<a name="1.12.1"></a>
## [1.12.1](https://github.com/IniZio/reim/compare/v1.12.0...v1.12.1) (2018-09-27)




**Note:** Version bump only for package reim

<a name="1.12.0"></a>
# [1.12.0](https://github.com/IniZio/reim/compare/v1.10.5...v1.12.0) (2018-09-25)


### Bug Fixes

* **typing:** transforms some implicit any types into explicity ([#32](https://github.com/IniZio/reim/issues/32)) ([f8c9ab5](https://github.com/IniZio/reim/commit/f8c9ab5))


### Features

* **diff:** use different lib for faster deep diff ([549340d](https://github.com/IniZio/reim/commit/549340d))
* **typing:** adds typescript and flow typing ([#31](https://github.com/IniZio/reim/issues/31)) ([f342ba4](https://github.com/IniZio/reim/commit/f342ba4)), closes [#29](https://github.com/IniZio/reim/issues/29)




<a name="1.11.0"></a>
# [1.11.0](https://github.com/IniZio/reim/compare/v1.10.5...v1.11.0) (2018-09-25)


### Bug Fixes

* **typing:** transforms some implicit any types into explicity ([#32](https://github.com/IniZio/reim/issues/32)) ([f8c9ab5](https://github.com/IniZio/reim/commit/f8c9ab5))


### Features

* **diff:** use different lib for faster deep diff ([549340d](https://github.com/IniZio/reim/commit/549340d))
* **typing:** adds typescript and flow typing ([#31](https://github.com/IniZio/reim/issues/31)) ([f342ba4](https://github.com/IniZio/reim/commit/f342ba4)), closes [#29](https://github.com/IniZio/reim/issues/29)




<a name="1.10.4"></a>
## [1.10.4](https://github.com/IniZio/reim/compare/v1.10.2...v1.10.4) (2018-08-29)


### Bug Fixes

* **react:** subscriber should update when children changes ([d4edf26](https://github.com/IniZio/reim/commit/d4edf26))




<a name="1.10.0"></a>
# [1.10.0](https://github.com/IniZio/reim/compare/v1.8.1...v1.10.0) (2018-08-23)


### Features

* **initial:** adds handling of initial value ([#26](https://github.com/IniZio/reim/issues/26)) ([355058e](https://github.com/IniZio/reim/commit/355058e))




<a name="1.9.0"></a>
# [1.9.0](https://github.com/IniZio/reim/compare/v1.8.1...v1.9.0) (2018-08-22)


### Features

* **initial:** adds handling of initial value ([97dc183](https://github.com/IniZio/reim/commit/97dc183))




<a name="1.8.0"></a>
# [1.8.0](https://github.com/IniZio/reim/compare/v1.7.0...v1.8.0) (2018-08-21)


### Features

* **react:** adds get/set method for store usage ([#23](https://github.com/IniZio/reim/issues/23)) ([a231533](https://github.com/IniZio/reim/commit/a231533))




<a name="1.7.0"></a>
# [1.7.0](https://github.com/IniZio/reim/compare/v1.6.0...v1.7.0) (2018-08-15)


### Features

* **reporter:** adds reporter plugin ([a257fb8](https://github.com/IniZio/reim/commit/a257fb8))
* **vue:** adds mutating `commit` method ([#20](https://github.com/IniZio/reim/issues/20)) ([a131d83](https://github.com/IniZio/reim/commit/a131d83))




<a name="1.6.0"></a>
# [1.6.0](https://github.com/IniZio/reim/compare/v1.5.6...v1.6.0) (2018-08-14)


### Features

* **plugin:** plugins can be added at runtime ([7829493](https://github.com/IniZio/reim/commit/7829493))
* **react:** adds pipeTo function to sync props to store e.g. react-router ([#17](https://github.com/IniZio/reim/issues/17)) ([1aeed44](https://github.com/IniZio/reim/commit/1aeed44))




<a name="1.5.6"></a>
## [1.5.6](https://github.com/IniZio/reim/compare/v1.5.5...v1.5.6) (2018-08-14)




**Note:** Version bump only for package reim

<a name="1.5.5"></a>
## [1.5.5](https://github.com/IniZio/reim/compare/v1.5.4...v1.5.5) (2018-08-14)

**Note:** Version bump only for package reim





<a name="1.5.4"></a>
## [1.5.4](https://github.com/IniZio/reim/compare/v1.5.1...v1.5.4) (2018-08-14)

**Note:** Version bump only for package reim





<a name="1.5.3"></a>
## [1.5.3](https://github.com/IniZio/reim/compare/v1.5.1...v1.5.3) (2018-08-14)

**Note:** Version bump only for package reim





<a name="1.5.2"></a>
## [1.5.2](https://github.com/IniZio/reim/compare/v1.5.1...v1.5.2) (2018-08-14)

**Note:** Version bump only for package reim





<a name="1.4.3"></a>
## [1.4.3](https://github.com/IniZio/reim/compare/v1.4.2...v1.4.3) (2018-08-13)

**Note:** Version bump only for package reim





<a name="1.4.2"></a>
## [1.4.2](https://github.com/IniZio/reim/compare/v1.4.0...v1.4.2) (2018-08-13)

**Note:** Version bump only for package reim





<a name="1.4.0"></a>
# [1.4.0](https://github.com/IniZio/reim/compare/v1.3.0...v1.4.0) (2018-08-13)


### Features

* **persist:** adds persistance plugin ([3b1e354](https://github.com/IniZio/reim/commit/3b1e354))





<a name="1.3.1"></a>
## [1.3.1](https://github.com/IniZio/reim/compare/v1.3.0...v1.3.1) (2018-08-13)

**Note:** Version bump only for package reim





<a name="1.3.0"></a>
# [1.3.0](https://github.com/IniZio/reim/compare/v1.2.0...v1.3.0) (2018-08-12)


### Features

* **observable:** makes reim store convertible to observable stream ([#11](https://github.com/IniZio/reim/issues/11)) ([a3428a9](https://github.com/IniZio/reim/commit/a3428a9)), closes [#10](https://github.com/IniZio/reim/issues/10)





<a name="1.2.0"></a>
# [1.2.0](https://github.com/IniZio/reim/compare/v1.0.1...v1.2.0) (2018-08-08)


### Features

* adds context and store api ([52848b2](https://github.com/IniZio/reim/commit/52848b2))





<a name="1.1.0"></a>
# [1.1.0](https://github.com/IniZio/reim/compare/v1.0.1...v1.1.0) (2018-08-07)


### Features

* adds context and store api ([52848b2](https://github.com/IniZio/reim/commit/52848b2))





<a name="1.0.1"></a>
## [1.0.1](https://github.com/IniZio/reim/compare/v1.0.0...v1.0.1) (2018-08-06)

**Note:** Version bump only for package reim





<a name="1.0.0"></a>
# [1.0.0](https://github.com/IniZio/reim/compare/v0.4.1...v1.0.0) (2018-08-06)


### Features

* adds setter prop for changing state ([544a4f8](https://github.com/IniZio/reim/commit/544a4f8))


### BREAKING CHANGES

* now uses setState instead of dispatch / commit





<a name="0.4.1"></a>
## [0.4.1](https://github.com/IniZio/reim/compare/v0.4.0...v0.4.1) (2018-08-04)


### Bug Fixes

* use getter to prevent this.state from modified from outside ([7948b1b](https://github.com/IniZio/reim/commit/7948b1b))





<a name="0.4.0"></a>
# [0.4.0](https://github.com/IniZio/reim/compare/v0.3.1...v0.4.0) (2018-08-04)


### Features

* **immuta:** uses immuta for support of Map and Set ([7ecc962](https://github.com/IniZio/reim/commit/7ecc962))





<a name="0.3.1"></a>
## [0.3.1](https://github.com/IniZio/reim/compare/v0.3.0...v0.3.1) (2018-08-03)

**Note:** Version bump only for package reim





<a name="0.3.0"></a>
# 0.3.0 (2018-08-02)


### Features

* **memoize:** memoizes dispatch and commit ([0f84832](https://github.com/IniZio/reim/commit/0f84832))
* **react:** adds react-reim for use with react ([7a7c879](https://github.com/IniZio/reim/commit/7a7c879))
