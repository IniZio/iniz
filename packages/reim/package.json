{
  "name": "reim",
  "version": "3.0.0",
  "license": "MIT",
  "sideEffects": false,
  "keywords": [
    "reim",
    "react",
    "store",
    "hoc",
    "immer"
  ],
  "main": "dist/reim.cjs.js",
  "module": "dist/reim.es.js",
  "unpkg": "dist/reim.js",
  "jsdelivr": "dist/reim.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "lib",
    "src"
  ],
  "scripts": {
    "start": "npm run dev",
    "dev": "bili --no-babel.babelrc --watch",
    "build": "npm run build:dist && cpx \"src/immer.{d.ts,js.flow}\" dist",
    "build:dist": "rimraf dist && bili --no-babel.babelrc"
  },
  "xo": {
    "parser": "babel-eslint",
    "extends": "xo-react",
    "semicolon": false,
    "esnext": true,
    "space": 2,
    "envs": [
      "node",
      "browser",
      "es6",
      "serviceworker"
    ],
    "globals": [
      "it"
    ],
    "rules": {
      "import/no-unassigned-import": 0,
      "capitalized-comments": 0,
      "quote-props": [
        "warn",
        "consistent"
      ],
      "react/button-has-type": 0
    }
  },
  "devDependencies": {
    "@types/auto-bind": "^1.2.0",
    "@types/eslint": "^4.16.6",
    "@types/event-emitter": "^0.3.3",
    "@types/lodash": "^4.14.123",
    "babel-eslint": "^8.2.6",
    "eslint": "^5.2.0",
    "eslint-config-xo-react": "^0.17.0",
    "eslint-plugin-react": "^7.10.0",
    "xo": "^0.21.1"
  },
  "dependencies": {
    "auto-bind": "^1.2.1",
    "event-emitter": "^0.3.5",
    "fast-deep-equal": "^2.0.1",
    "immer": "8",
    "is-plain-object": "^3.0.0"
  },
  "gitHead": "419591f08507517258271f77077dc7a1d5c1f7af"
}
