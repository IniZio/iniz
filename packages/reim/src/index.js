import {produce} from 'immer'
import isEqual from 'lodash/isEqual'

export class Store {
  state = null

  _subscribers = []

  constructor(state) {
    this.state = produce(null, () => state)
  }

  /**
   * dispatch - Execute effect
   *
   * @param {Function} effect An effect has side-effect
   *
   * @returns {Function} dispatched effect
   */
  dispatch(effect) {
    return async (...args) => {
      const m = await effect(...args)
      const mutations = [].concat(m)
      mutations.map((...args) => this.commit(...args)())
    }
  }

  /**
   * commit - Execute mutations
   *
   * @param {Function} mutations Mutations only mutate state
   *
   * @returns {any} New state
   */
  commit(mutations) {
    return (...args) => {
      this.state = produce(this.state, draft => ([].concat(mutations)).forEach(mutation => mutation(draft, ...args)))
      this._notify()
      return this.state
    }
  }

  /**
   * subscribe - Add subscriber
   *
   * @param {Function} handler Subscriber callback
   * @param {Object} option Options for subscription
   * @param {boolean} option.immediate Whether to run sub
   * @param {Function} option.selector Selectors for data subscription
   *
   * @returns {Function} Unsubscriber
   */
  subscribe(handler, {immediate = true, selector = state => state} = {}) {
    this._subscribers.push({handler, selector})
    if (immediate) {
      this._notify()
    }
    return () => this.unsubscribe(handler)
  }

  /**
   * unsubscribe - Remove subscriber
   *
   * @param {Function} handler Subscriber to remvoe
   *
   */
  unsubscribe(handler) {
    this._subscribers.splice(this._subscribers.findIndex(sub => sub.handler === handler), 1)
  }

  _notify() {
    this._subscribers.forEach(sub => {
      // Notify if selected is updated
      const selected = produce(this.state, sub.selector)
      if (!isEqual(selected, sub.selected)) {
        sub.selected = selected
        sub.handler(selected)
      }
    })
  }
}

export const register = state => new Store(state)
