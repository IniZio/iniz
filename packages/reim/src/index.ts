import {isPlainObject, isFunction} from 'lodash'
import produce from 'immer'
const equal = require('fast-deep-equal')
const autoBind = require('auto-bind')

import {Actions, ReimOptions, Action, Handler, Filter, SnapshotFor, Meta} from './types'

const isReimFlag = Symbol('reim')

const instances: {[name: string]: any} = {}

// @ts-ignore
const observableSymbol = () => ((typeof Symbol === 'function' && Symbol.observable) || '@@observable')

const withDevTools = (
  // @ts-ignore
  process.env.NODE_ENV === 'development' &&
  // @ts-ignore
  typeof window !== 'undefined' && window.__REDUX_DEVTOOLS_EXTENSION__
)

export function isReim(store: any): store is Reim {
  return store.__isReim === isReimFlag
}

export class Reim<T = any> {
  _name?: string
  _initial: T
  _state: T
  _subscribers: {handler: Handler<T>, filter: Filter<T>, cache: any}[] = []
  _devTools: any

  get __isReim() { return isReimFlag }
  get name() { return this._name }

  constructor(initial?: T, options: ReimOptions<T> = {}) {
    autoBind(this)

    this._name = options.name
    this._state = initial

    if (this.name) {
      if (instances[this.name]) {
        throw new Error(`There is already an instance named ${this._name}, named stores must have unique names`)
      }
      instances[this.name] = {}
    }

    if (withDevTools) {
      // @ts-ignore
      this._devTools = window.__REDUX_DEVTOOLS_EXTENSION__.connect({instanceId: this.name, shouldStringify: false})
      this._devTools.subscribe((message: any) => {
        if (message.id !== this._name) {
          return
        }

        if (message.type === 'DISPATCH' && (message.payload.type === 'JUMP_TO_STATE' || message.payload.type === 'JUMP_TO_ACTION')) {
          this.reset(typeof message.state === 'string' ? JSON.parse(message.state) : message.state)
        }
      })
      this._devTools.init(this.filter())
    }

    Object.assign(this, this.actions(options.actions))
  }

  filter<
    TF extends Filter<T>
  >(filter?: TF): TF extends (null | undefined) ? T : SnapshotFor<TF, T> {
    if (!filter) {
      return this._state as TF extends (null | undefined) ? T : SnapshotFor<TF, T>
    }

    if (this._state[filter as keyof T]) {
      return this._state[filter as keyof T] as TF extends (null | undefined) ? T : SnapshotFor<TF, T>
    }

    if (typeof filter === 'function') {
      return (filter as (s: T) => any)(this._state)
    }

    if (isPlainObject(filter)) {
      return Object.entries(filter as {[index: string]: ((s: T) => any)})
        .reduce((acc, [key, f]) => ({
          ...acc,
          [key]: (f as (s: T) => any)(this._state)
        }), {} as TF extends (null | undefined) ? T : SnapshotFor<TF, T>)
    }

    return this._state as TF extends (null | undefined) ? T : SnapshotFor<TF, T>
  }

  set<T, TR extends Reim<T>>(this: TR & {_state: T}, action: Action<T>, {mutation, payload}: Meta = {}) {
    const _mutation = action(...payload)

    this._state = isPlainObject(this._state) ? (
      produce(
        this._state, (
          isFunction(_mutation) ?
            state => _mutation(state) :
            state => void Object.assign(state, _mutation)
        )
      )
    ) : (
      isFunction(_mutation) ?
        _mutation(this._state) as T :
        _mutation as T
    )

    this._notify({mutation, payload})
  }

  _notify = (meta: {mutation: string, payload: any[]}) => {
    this._subscribers.forEach(sub => {
      // Notify if cache is updated
      const cache = this.filter(sub.filter)

      if (!equal(cache, sub.cache)) {
        sub.cache = cache;
        (isFunction(sub.handler) ? sub.handler : sub.handler.next)(cache, meta)
      }
    })
  }

  reset<T, TR extends Reim<T>>(this: TR & {_state: T}, initial: T = null) {
    if (initial) {
      this._initial = initial
    }

    this.set(() => this._initial)
    return this.filter()
  }

  actions<
    TR extends Reim<T>,
    T,
    TA extends Actions<T>
  >(this: TR & {_state: T}, actions: TA = {} as TA): {[k in keyof typeof actions]: TA[k]} {
    return Object.keys(actions)
      .reduce((acc, key) => ({
        ...acc,
        [key]: (...args: any[]) => void (
          this.set(actions[key], {mutation: key, payload: args})
        )
      }), {}) as {[k in keyof typeof actions]: TA[k]}
  }

  unsubscribe<TR extends Reim<T>, T>(this: TR & {_state: T}, handler: Handler<T>) {
    this._subscribers.splice(this._subscribers.findIndex(sub => sub.handler === handler), 1)
  }

  subscribe<TR extends Reim<T>, T>(this: TR & {_state: T}, handler: Handler<T>, {immediate = false, filter}: {immediate?: boolean; filter?: Filter<T>} = {}) {
    // TODO: check if the filter is not cachable here
    const cache = this.filter(filter)

    this._subscribers.push({handler, filter, cache})
    if (immediate) {
      (isFunction(handler) ? handler : handler.next)(cache)
    }
    return {unsubscribe: () => this.unsubscribe(handler)}
  }
  [observableSymbol()]() {
    return this
  }
}

const reim = <T, TA>(initial: T, options: ReimOptions<T> & {actions?: TA} = {}): Reim & {[k in keyof TA]: TA[k]} => {
  const instance = new Reim<T | null | undefined>(initial, options)

  return Object.assign(instance, instance.actions(options.actions))
}

export * from './types'

export default reim
