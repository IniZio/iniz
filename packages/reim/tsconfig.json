{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "declarationDir": "./dist",
    "typeRoots" : ["./typings"]
  },
  "include": [
    "./src",
    "./typings"
  ]
}
