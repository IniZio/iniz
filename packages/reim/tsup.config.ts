import { defineConfig } from 'tsup'

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['iife', 'esm', 'cjs'],
  target: 'es5',
  treeshake: true,
  splitting: false,
  sourcemap: true,
  clean: true,
  dts: true
})