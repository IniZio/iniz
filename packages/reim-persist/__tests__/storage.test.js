import Storage from 'dom-storage'
import {store} from '../../reim'
import persist from '../src'

test('can be created', () => {
  const tstore = store({
    abc: 12
  }, {
    name: 'xyz',
    plugins: [
      persist()
    ]
  })

  tstore.setState(state => {
    state.abc += 100
  })

  expect(tstore.state.abc).toBe(112)
})

test('replace store state and should subscribe to change when initialized', () => {
  const storage = new Storage()
  storage.setItem('reim/abc', JSON.stringify({persisted: 'json'}))

  const st = store({
    xx: 234
  }, {
    name: 'abc'
  })

  st.subscribe = jest.fn()

  const plugin = persist({storage})

  plugin.call(st)

  expect(st.state).toEqual({
    xx: 234,
    persisted: 'json'
  })
  expect(st.subscribe).toBeCalled()
})

test('should not replace store state when saved state is invalid', () => {
  const storage = new Storage()
  storage.setItem('reim/abc', 'invalid')

  const st = store({
    xx: 234
  }, {name: 'abc'})

  st.subscribe = jest.fn()
  st.setState = jest.fn()

  const plugin = persist({storage})

  plugin.call(st)

  expect(st.setState).not.toBeCalled()
  expect(st.subscribe).toBeCalled()
})

test('should merge saved and current store recursively and replace values', () => {
  const storage = new Storage()
  storage.setItem('reim/abc', JSON.stringify({xx: [12, 33], zz: {ee: 38}}))

  const st = store({
    xx: [234],
    y: 'as',
    zz: {
      mag: 11
    }
  }, {name: 'abc'})

  st.subscribe = jest.fn()

  const plugin = persist({storage})

  plugin.call(st)

  expect(st.state).toEqual({xx: [12, 33], y: 'as', zz: {mag: 11, ee: 38}})

  expect(st.subscribe).toBeCalled()
})
