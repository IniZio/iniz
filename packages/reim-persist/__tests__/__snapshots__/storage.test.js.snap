// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should report anonymous store trying to use persist plugin 1`] = `"You cannot persist an anonymous store, use \`name\` option in store"`;

exports[`should report invalid storage 1`] = `"Invalid storage given"`;
